File: source/algorithms/src/simd.c
#include "simd.h"

#define SIMD_SIZE 128
#define SIMD_SIZE_BYTE SIMD_SIZE / 8

void add_arrays_u8_u8_u16(const uint8_t* a, const uint8_t* b, uint16_t* c, size_t size) {

    size_t i = 0;
    size_t step_size = SIMD_SIZE_BYTE / sizeof(uint16_t);

    for (; i < (size - step_size); i += step_size) {
        // Load 8 uint8_t elements from array1 and array2
        uint8x8_t neon1 = vld1_u8(a + i);
        uint8x8_t neon2 = vld1_u8(b + i);

        // Convert uint8_t to uint16_t by zero-extending and then add
        uint16x8_t sum = vaddl_u8(neon1, neon2);

        // Store the result in result array as uint16_t
        vst1q_u16(c + i, sum);
    }

    for (; i < size; i++){
        c[i] = (uint16_t) a[i] + b[i];
    }
    
}

void add_arrays_u16_u16_u16(const uint16_t* a, const uint16_t* b, uint16_t* c, size_t size) {

    size_t i = 0;
    size_t step_size = SIMD_SIZE_BYTE / sizeof(uint16_t);

    for (; i < (size - step_size); i += step_size) {
        // Load 8 uint8_t elements from array1 and array2
        uint16x8_t neon1 = vld1q_u16(a + i);
        uint16x8_t neon2 = vld1q_u16(b + i);

        // Convert uint8_t to uint16_t by zero-extending and then add
        uint16x8_t sum = vaddq_u16(neon1, neon2);

        // Store the result in result array as uint16_t
        vst1q_u16(c + i, sum);
    }

    for (; i < size; i++){
        c[i] = a[i] + b[i];
    }
    
}

void add_arrays_u16_u16_u32(const uint16_t* a, const uint16_t* b, uint32_t* c, size_t size) {

    size_t i = 0;
    size_t step_size = SIMD_SIZE_BYTE / sizeof(uint32_t);

    for (; i < (size - step_size); i += step_size) {
        // Load 8 uint8_t elements from array1 and array2
        uint16x4_t neon1 = vld1_u16(a + i);
        uint16x4_t neon2 = vld1_u16(b + i);

        // Convert uint8_t to uint16_t by zero-extending and then add
        uint32x4_t sum = vaddl_u16(neon1, neon2);

        // Store the result in result array as uint16_t
        vst1q_u32(c + i, sum);
    }

    for (; i < size; i++){
        c[i] = (uint32_t) a[i] + b[i];
    }
    
}

File: source/algorithms/src/simd.rs
use rayon::prelude::*;
use std::arch::aarch64::*;

const SIMD_SIZE_BYTE: usize = 16;
// Function to sum two slices of u8 into a SIMD vector of u16
pub fn add_slices_u8_u8_u16(a: &[u8], b: &[u8]) -> Vec<u16> {
    assert_eq!(a.len(), b.len(), "a and b are not in the same length");

    let size: usize = a.len();
    let step_size = SIMD_SIZE_BYTE / std::mem::size_of::<u16>(); // Size in elements of uint16x8_t
    let last_index = (size / step_size) * step_size;

    let mut c = vec![0; size];

    for i in (0..size).step_by(step_size) {
        unsafe {
            // Load 8 uint8_t elements from array1 and array2
            let neon1 = vld1_u8(a.as_ptr().add(i));
            let neon2 = vld1_u8(b.as_ptr().add(i));
            let sum = vaddl_u8(neon1, neon2);
            vst1q_u16(c.as_mut_ptr().add(i), sum);
        }
    }

    for i in last_index..size {
        c[i] = a[i] as u16 + b[i] as u16;
    }

    return c;
}
